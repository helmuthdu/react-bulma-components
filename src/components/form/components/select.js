import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Select = ({
  children,
  className,
  color,
  disabled,
  empty,
  loading,
  multiple,
  name,
  readOnly,
  rounded,
  size,
  style,
  value,
  ...allProps
}) => {
  const props = modifiers.clean(allProps);
  return (
    <div
      className={cn('select', modifiers.getClassName(allProps), className, {
        [`is-${size}`]: size,
        [`is-${color}`]: color,
        'is-loading': loading,
        'is-rounded': rounded,
        'is-empty': empty,
        'is-multiple': multiple
      })}
      style={style}
    >
      <select {...props} multiple={multiple} value={value} aria-readonly={readOnly} disabled={disabled} name={name}>
        {children}
      </select>
    </div>
  );
};

Select.propTypes = {
  ...modifiers.propTypes,
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  empty: PropTypes.bool,
  loading: PropTypes.bool,
  multiple: PropTypes.bool,
  name: PropTypes.string,
  readOnly: PropTypes.bool,
  rounded: PropTypes.bool,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

Select.defaultProps = {
  ...modifiers.defaultProps,
  empty: false,
  loading: false,
  multiple: false,
  rounded: false,
  value: ''
};
