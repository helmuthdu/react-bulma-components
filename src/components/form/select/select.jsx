import cn from 'classnames';
import PropTypes from 'prop-types';
import React from 'react';
import CONSTANTS from '../../../constants';
import modifiers from '../../../modifiers';

const colors = Object.values(CONSTANTS.COLORS);

export const Select = ({
  children,
  className,
  color,
  disabled,
  empty,
  loading,
  multiple,
  name,
  readOnly,
  rounded,
  size,
  style,
  value,
  ...allProps
}) => {
  const props = modifiers.clean(allProps);
  return (
    <div
      className={cn('select', modifiers.getClassName(allProps), className, {
        'is-empty': empty,
        'is-loading': loading,
        'is-multiple': multiple,
        'is-rounded': rounded,
        [`is-${color}`]: color,
        [`is-${size}`]: size
      })}
      style={style}
    >
      <select {...props} multiple={multiple} value={value} aria-readonly={readOnly} disabled={disabled} name={name}>
        {children}
      </select>
    </div>
  );
};

Select.propTypes = {
  ...modifiers.propTypes,
  children: PropTypes.node,
  className: PropTypes.string,
  color: PropTypes.oneOf(colors),
  disabled: PropTypes.bool,
  empty: PropTypes.bool,
  loading: PropTypes.bool,
  multiple: PropTypes.bool,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  readOnly: PropTypes.bool,
  rounded: PropTypes.bool,
  size: PropTypes.oneOf(Object.values(CONSTANTS.SIZES)),
  style: PropTypes.object,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number, PropTypes.array])
};

Select.defaultProps = {
  ...modifiers.defaultProps,
  empty: false,
  loading: false,
  multiple: false,
  rounded: false,
  value: ''
};
