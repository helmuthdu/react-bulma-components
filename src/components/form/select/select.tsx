import cn from 'classnames';
import React from 'react';
import { Colors, Sizes } from '../../../constants';
import modifiers, { Modifiers } from '../../../modifiers';

type SelectProps = Partial<Modifiers> & {
  color?: Colors;
  disabled?: boolean;
  empty?: boolean;
  loading?: boolean;
  multiple?: boolean;
  name?: string;
  onChange: (...args: any[]) => any;
  readOnly?: boolean;
  rounded?: boolean;
  size?: Sizes;
  value?: string | number | any[];
};

export const Select: React.FunctionComponent<SelectProps> = ({
  children,
  className,
  color,
  disabled,
  empty,
  loading,
  multiple,
  name,
  readOnly,
  rounded,
  size,
  style,
  value,
  ...allProps
}: SelectProps) => {
  const props = modifiers.clean(allProps);
  return (
    <div
      className={cn('select', modifiers.getClassName(allProps), className, {
        'is-empty': empty,
        'is-loading': loading,
        'is-multiple': multiple,
        'is-rounded': rounded,
        [`is-${color}`]: color,
        [`is-${size}`]: size
      })}
      style={style}
    >
      <select {...props} multiple={multiple} value={value} aria-readonly={readOnly} disabled={disabled} name={name}>
        {children}
      </select>
    </div>
  );
};

Select.defaultProps = {
  ...modifiers.defaultProps,
  empty: false,
  loading: false,
  multiple: false,
  rounded: false,
  value: ''
};
