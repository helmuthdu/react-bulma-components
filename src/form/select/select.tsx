import clsx from 'clsx';
import * as React from 'react';
import { ChangeEvent, useEffect, useState } from 'react';
import { Colors, Sizes } from '../../constants';
import modifiers, { ElementModifier } from '../../modifiers';

type SelectProps = ElementModifier & {
  color?: Colors;
  disabled?: boolean;
  empty?: boolean;
  loading?: boolean;
  multiple?: boolean;
  readOnly?: boolean;
  rounded?: boolean;
  size?: Sizes;
  value?: string | number | any[];
} & Omit<React.ComponentProps<'select'>, 'ref' | 'size' | 'color' | 'unselectable'>;

export const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
  (
    {
      children,
      className,
      color,
      disabled,
      empty,
      loading,
      multiple,
      name,
      readOnly,
      rounded,
      size,
      style,
      value: initialValue,
      onChange,
      ...props
    },
    ref
  ) => {
    const [value, setValue] = useState(initialValue);

    useEffect(() => {
      setValue(initialValue);
    }, [initialValue]);

    return (
      <div
        className={clsx('select', modifiers.getClassName(props), className, {
          'is-empty': empty,
          'is-loading': loading,
          'is-multiple': multiple,
          'is-rounded': rounded,
          [`is-${color}`]: color,
          [`is-${size}`]: size
        })}
        style={style}
      >
        <select
          ref={ref}
          aria-readonly={readOnly}
          disabled={disabled}
          multiple={multiple}
          name={name}
          value={value}
          onChange={(evt: ChangeEvent<HTMLSelectElement>) => {
            setValue(evt.target.value);
            if (onChange) onChange(evt);
          }}
          {...modifiers.clean(props)}
        >
          {children}
        </select>
      </div>
    );
  }
);

Select.defaultProps = {
  ...modifiers.defaultProps,
  empty: false,
  loading: false,
  multiple: false,
  rounded: false,
  value: ''
};
