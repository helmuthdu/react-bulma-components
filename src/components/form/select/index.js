import './select.sass';

export * from './select';
