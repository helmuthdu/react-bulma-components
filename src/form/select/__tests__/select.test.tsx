import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';
import { Select } from '../select';

describe('Select component', () => {
  let onChange: jest.Mock;

  beforeEach(() => {
    onChange = jest.fn();
  });

  it('should render', () => {
    const { asFragment } = render(
      <Select onChange={onChange}>
        <option>1</option>
      </Select>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should concat classname in props with classname', () => {
    const { container } = render(
      <Select onChange={onChange} className="other-class this-is-a-test">
        <option>1</option>
        <option>2</option>
        <option>3</option>
      </Select>
    );
    const select = container.querySelector('.select') as HTMLSelectElement;
    expect(select).toHaveClass('other-class', 'this-is-a-test');
  });

  it('should use inline styles', async () => {
    const { container } = render(
      <Select onChange={onChange} style={{ width: '100%' }}>
        <option>1</option>
        <option>2</option>
        <option>3</option>
      </Select>
    );
    const select = container.querySelector('select') as HTMLSelectElement;
    fireEvent.change(select, { target: { value: 2 } });
    expect(select).toHaveValue('2');
  });

  it('should be large, red, disabled and multioption', () => {
    const { container } = render(
      <Select onChange={onChange} value={[]} color="danger" size="large" multiple disabled>
        <option>1</option>
        <option>2</option>
        <option>3</option>
      </Select>
    );
    const select = container.querySelector('select') as HTMLSelectElement;
    fireEvent.change(select, { target: { value: [3] } });
    expect(select).toHaveValue(['3']);
  });
});
