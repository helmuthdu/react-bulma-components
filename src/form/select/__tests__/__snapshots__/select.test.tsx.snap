// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render 1`] = `
<DocumentFragment>
  <div
    class="select"
  >
    <select>
      <option>
        1
      </option>
    </select>
  </div>
</DocumentFragment>
`;
